File: src/FSOps.Core/Entities/Contract.cs
namespace FSOps.Core.Entities;

/// <summary>
/// One job offered by another operator: the player flies it personally, in an aircraft the operator
/// supplies, and is paid a fee. <b>They bear no operating costs</b> - fuel, landing, handling and
/// maintenance all belong to the other business. That is the whole point of the arrangement and it
/// is enforced structurally rather than by a rule: the contract completion path never reaches the
/// code that would post those lines (see FlightLifecycleService.FinalizeFlightAsync).
///
/// <para><b>This is not a Route and must never become one.</b> A Route is the player's own airline
/// selling seats between two airports it serves, and it carries a fare, a demand model and a flight
/// number. A contract is somebody else's aeroplane going somewhere for somebody else's reasons. They
/// are stored separately for the same reason the fleet catalogue and the contract aircraft list are
/// separate: merging them would quietly feed jobs flown for other operators into the demand, seat
/// and reputation models that exist to describe the player's own airline.</para>
///
/// <para><b>The stops are named up front.</b> A ferry lists its whole chain - Bristol, Wick,
/// Reykjavik, Narsarsuaq, Goose Bay, New York - and every one of those legs was checked against this
/// aircraft's range when the contract was generated. A job on the board is always flyable by the
/// aircraft it names, and the player always knows exactly what they accepted before they accept it.
/// </para>
/// </summary>
public class Contract
{
    public Guid Id { get; set; }

    public Guid AirlineId { get; set; }

    public ContractKind Kind { get; set; }

    public ContractStatus Status { get; set; } = ContractStatus.Offered;

    /// <summary>
    /// The board this offer belongs to - see <c>ContractBoardKey</c>. Contracts are generated
    /// deterministically per (world seed, airline, time bucket), and this records which bucket
    /// produced this row so the board can be regenerated idempotently: a second read of the same
    /// bucket returns the rows already stored rather than a second set of near-identical offers.
    /// It also identifies which offers a later bucket has moved past, so they expire rather than
    /// accumulating for ever.
    /// </summary>
    public long BoardBucket { get; set; }

    /// <summary>
    /// A stable index within the board bucket, 0-based. With <see cref="BoardBucket"/> this is what
    /// makes an offer identifiable across regenerations of the same board.
    /// </summary>
    public int BoardSlot { get; set; }

    /// <summary>The other business, as a name to show. Generated, and purely flavour - nothing keys off it.</summary>
    public string OperatorName { get; set; } = string.Empty;

    /// <summary>
    /// The ICAO type designator of the aircraft the contract supplies - a key into
    /// <see cref="FSOps.Core.SimAircraft.ContractAircraftCatalogue"/>, never into the airline's own
    /// fleet or into AircraftTypes. Stored as the designator rather than as a foreign key precisely
    /// because there is nothing in the database to point at: the aeroplane belongs to somebody else.
    /// </summary>
    public string AircraftTypeDesignator { get; set; } = string.Empty;

    /// <summary>What the job is carrying, as words: "2,400 kg of machine parts", "4 passengers", or empty for a ferry.</summary>
    public string LoadDescription { get; set; } = string.Empty;

    /// <summary>Freight, kilograms. Zero for a ferry (the aeroplane is the cargo) and for a charter.</summary>
    public int PayloadKg { get; set; }

    /// <summary>Passengers. Zero for a ferry and for cargo.</summary>
    public int PaxCount { get; set; }

    /// <summary>
    /// The whole job's fee, in the stored base unit, for flying every leg. Never paid as a lump:
    /// each leg carries its own share (see <see cref="ContractLeg.FeeShare"/>), those shares sum
    /// exactly to this, and each is posted as its leg is actually flown.
    /// </summary>
    public decimal Fee { get; set; }

    public double TotalDistanceNm { get; set; }

    public int TotalPlannedBlockMinutes { get; set; }

    /// <summary>When this offer appeared on the board.</summary>
    public DateTimeOffset OfferedUtc { get; set; }

    /// <summary>
    /// When the job must be finished by. <b>Fixed when the contract is generated and visible before
    /// the player accepts</b> - it never ambushes them, and the world stays predictable. Generous by
    /// design: weeks of real time, so a half-finished ocean crossing can be left and picked up.
    /// </summary>
    public DateTimeOffset DeadlineUtc { get; set; }

    public DateTimeOffset? AcceptedUtc { get; set; }

    /// <summary>When this contract reached a terminal status - completed, abandoned or expired.</summary>
    public DateTimeOffset? ClosedUtc { get; set; }

    /// <summary>
    /// Why an abandoned contract ended, in words the player can read: they gave up, or the deadline
    /// passed. Null for every other status. A contract that silently stopped being available is the
    /// worst possible outcome, so the reason is always recorded rather than inferred from dates.
    /// </summary>
    public string? ClosedReason { get; set; }

    public DateTimeOffset CreatedUtc { get; set; }

    public DateTimeOffset? DeletedUtc { get; set; }

    public List<ContractLeg> Legs { get; set; } = new();
}

/// <summary>
/// One sector of a contract, flown in order. A ferry has several; cargo and charter normally have
/// one, but nothing forbids more - leg count falls out of distance against the named aircraft's
/// range rather than being decided by the kind.
/// </summary>
public class ContractLeg
{
    public Guid Id { get; set; }

    public Guid ContractId { get; set; }

    /// <summary>1-based position in the chain. Legs are flown in this order and no other.</summary>
    public int Sequence { get; set; }

    public string DepartureIcao { get; set; } = string.Empty;

    public string ArrivalIcao { get; set; } = string.Empty;

    public double DistanceNm { get; set; }

    /// <summary>
    /// What this leg was expected to take, from the contract aircraft's cruise speed. This is the
    /// weight in <see cref="FeeShare"/> and in the abandon charge, and it is stamped at generation
    /// so neither can move under the player after they accepted.
    /// </summary>
    public int PlannedBlockMinutes { get; set; }

    /// <summary>
    /// This leg's slice of <see cref="Contract.Fee"/>, weighted by
    /// <see cref="PlannedBlockMinutes"/>. Stamped at generation, so what a leg is worth is a fact
    /// about the contract rather than something recomputed (and therefore able to differ) at the
    /// moment it is paid. Every leg's share sums exactly to the fee - see
    /// <see cref="FSOps.Core.Contracts.ContractPayCalculator.AllocateFeeShares"/>.
    /// </summary>
    public decimal FeeShare { get; set; }

    /// <summary>
    /// The flight that flew this leg, once one has. Null while the leg is outstanding. This is the
    /// only link between a contract and a Flight row, and it is the thing that makes "pay per leg
    /// actually flown" answerable from stored fact rather than from a running tally.
    /// </summary>
    public Guid? FlightId { get; set; }

    public DateTimeOffset? FlownUtc { get; set; }

    public DateTimeOffset? DeletedUtc { get; set; }
}

File: src/fsops-web/src/types/contract.ts
/**
 * Contract flying - jobs other operators offer, flown personally by the player in the operator's own
 * aircraft for a flat fee.
 *
 * Every shape here mirrors what `ContractEndpoints` actually returns. Two things about that are
 * deliberate and worth not undoing:
 *
 * - **The server does the ordering, the arithmetic and the refusing.** A contract carries its whole
 *   chain of stops, the per-leg fee shares, what has been earned so far, what is outstanding, and
 *   which leg is next. None of that is re-derived in the client, so the board can never disagree
 *   with the ledger about which sector comes next or what it pays.
 * - **The sentences are the API's, not the UI's.** `limitation.message`, `abandonReason`,
 *   `closedReason` and every `error` are written to be read by a player and are rendered verbatim.
 *   Rewording them here is how two screens come to say different things about one number.
 */

/** Which kind of job. Ferry is the multi-leg expedition; the other two exist so the board is not all
 *  one shape. See `ContractKind` in Enums.cs. */
export type ContractKind = 'Ferry' | 'Cargo' | 'Charter'

/**
 * Where a contract is in its life (`ContractStatus` in Enums.cs).
 *
 * `Expired` and `Abandoned` are deliberately distinct and must not be conflated in display: an offer
 * nobody took costs nothing, whereas a job walked out of half-way carries a charge.
 */
export type ContractStatus = 'Offered' | 'Accepted' | 'Completed' | 'Abandoned' | 'Expired'

/** The aeroplane the job names, from the contract aircraft catalogue. Fields are nullable because a
 *  catalogue entry can be retired without invalidating contracts already written against it - render
 *  a null as "not known", never as zero. */
export interface ContractAircraftInfo {
  typeDesignator: string
  name: string | null
  manufacturer: string | null
  /** `ContractAircraftCategory.ToString()` - display only, nothing branches on it. */
  category: string | null
  rangeNm: number | null
  cruiseTasKts: number | null
  seats: number | null
}

/** One sector of a contract. `feeShare` is stamped when the contract is generated and never
 *  recomputed, which is why the earned/outstanding split can always agree with the ledger. */
export interface ContractLeg {
  id: string
  /** 1-based position in the chain. Legs are flown in this order and the server enforces it. */
  sequence: number
  departureIcao: string
  arrivalIcao: string
  distanceNm: number
  plannedBlockMinutes: number
  /** What this leg is WORTH - stamped when the contract was generated. */
  feeShare: number
  /**
   * What this leg actually PAID, from the posted ledger rows.
   *
   * Equal to `feeShare` for a leg flown in the simulator; **zero** for one completed with estimates
   * or invalidated by slew or a position jump, both of which count as flown and pay nothing. `null`
   * when the leg has not been flown at all — a third state, not a payment of zero.
   */
  feePaid: number | null
  flown: boolean
  flightId: string | null
  flownUtc: string | null
}

/** The next sector to fly. Null when every leg has been flown. Supplied by the server precisely so a
 *  screen never has to work out the ordering rule for itself. */
export interface ContractNextLeg {
  id: string
  sequence: number
  departureIcao: string
  arrivalIcao: string
}

/** One contract, as the board and the job view render it (`ContractEndpoints.ToDto`). */
export interface Contract {
  id: string
  kind: ContractKind
  status: ContractStatus
  operatorName: string

  aircraft: ContractAircraftInfo

  /** A human sentence describing the load - "2,400 kg of machine parts", "6 passengers", or for a
   *  ferry "Positioning flight - Cessna 172, empty". Rendered as written. */
  loadDescription: string
  payloadKg: number
  paxCount: number

  /** The whole job's fee for flying every leg. What is actually banked follows the legs flown, not
   *  this figure. Does NOT include the completion bonus. */
  fee: number
  /**
   * A single lump paid **only when every leg has been flown**, on top of the per-leg shares. Zero for
   * a single-leg job, and it grows with the length of the chain.
   *
   * Shown before accepting on purpose: a bonus nobody knows about cannot influence the decision it
   * exists to influence. **Forfeited by handing the job back** — and it is never part of the abandon
   * charge, so walking away loses it rather than being billed for it.
   */
  completionBonus: number
  /** `fee + completionBonus` — what the job is worth if it is seen through, which is the figure a
   *  player actually compares between offers. */
  totalIfCompleted: number
  totalDistanceNm: number
  totalPlannedBlockMinutes: number

  legCount: number
  flownLegCount: number
  /**
   * What has actually been **banked**, summed from the posted ledger rows — not from the stamped
   * shares of legs that happen to be marked flown.
   *
   * Those are different numbers: a leg completed with estimates counts as flown and pays nothing, as
   * does one invalidated by slew or a position jump. Compare against `legs[].feeShare` to see what a
   * flown leg was worth, and `legs[].feePaid` to see what it paid.
   */
  earnedSoFar: number
  /** What the REMAINING legs are worth — a fact about the future, so still from the stamped shares. */
  outstandingFee: number

  /**
   * What handing this job back would cost right now, and the sentence explaining it - both straight
   * from `ContractPayCalculator.CalculateAbandonCharge` via the same call that posts the charge.
   *
   * This is on the DTO rather than worked out here for a specific reason: the charge is the
   * outstanding fee scaled by `ContractConfig.AbandonChargeFraction`, which is server-side economy
   * config the client cannot see. A client that multiplied `outstandingFee` itself would be right
   * only until that number was tuned, and would then quietly describe something other than what
   * happens. Show `abandonCharge`; never recompute it.
   *
   * Zero is a real answer with a real reason ("no leg was flown, so ... costs nothing"), not a
   * missing one.
   */
  abandonCharge: number
  abandonReason: string

  offeredUtc: string
  /** Visible before accepting and never recalculated afterwards, so it cannot move under the player. */
  deadlineUtc: string
  acceptedUtc: string | null
  closedUtc: string | null
  closedReason: string | null

  nextLeg: ContractNextLeg | null
  /** Present whenever the server was asked for the full chain, which the board and job view both are. */
  legs: ContractLeg[] | null
}

/**
 * Why a board is thinner than it should be, in terms the player can act on.
 *
 * `message` is null when the board came out full, and is deliberately never an empty string -
 * "nothing to say" and "something to say, and it is blank" must not look alike. When it IS set it
 * has to be shown: a thin board with no explanation is indistinguishable from a broken feature, and
 * both real causes (too few aircraft ticked, an airline that barely touches anywhere) are one click
 * from being fixed.
 */
export interface ContractBoardLimitation {
  availableAircraftCount: number
  originCount: number
  requested: number
  generated: number
  message: string | null
}

/** GET /contracts/board. */
export interface ContractBoard {
  bucket: number
  refreshesUtc: string
  offered: Contract[]
  /** Jobs the player has taken and not yet finished. These survive every board refresh. */
  accepted: Contract[]
  limitation: ContractBoardLimitation
}

/** POST /contracts/{id}/abandon - the charge actually raised, and the server's own words for it. */
export interface ContractAbandonResult {
  contract: Contract
  charge: number
  unflownLegCount: number
  unflownBlockMinutes: number
  reason: string
}

/** POST /contracts/{id}/start-leg - the leg that just became a real, tracked flight. */
export interface ContractStartLegResult {
  flightId: string
  contractId: string
  contractLegId: string
  legSequence: number
  legCount: number
  departureIcao: string
  arrivalIcao: string
  distanceNm: number
  plannedBlockMinutes: number
  feeShare: number
  aircraftTypeDesignator: string
  aircraftName: string | null
  /** Informational only and never penalised, exactly as for an airline sector. Null means the sim
   *  reported no aircraft to compare against - a third state, not a failed comparison. */
  typeMismatch: boolean | null
}

/**
 * The contract marker carried by a logbook row and a flight's report card. Null for an ordinary
 * airline sector, which is how a consumer tells the two apart without inferring it from a missing
 * `routeId`.
 */
export interface FlightContractInfo {
  contractId: string
  kind: ContractKind
  operatorName: string
  legSequence: number
  legCount: number
  feeShare: number
  /** Report card only - the logbook's row does not carry these. */
  status?: ContractStatus
  aircraftTypeDesignator?: string
  aircraftName?: string | null
  departureIcao?: string
  arrivalIcao?: string
}
